Add introduction to Oak Functions readme (#2367)
project-oak/oak: oak_functions/README.md

preserving way. Oak Functions leverages TEEs and remote attestation, Wasm
sandboxing, and allows exposing metrics in a Differentially Private way.

> At its core, the Oak Functions loader is a trusted runtime compiled to a server

Reviewer: I agree. Maybe just drop "loader"?
```suggestion
At its core, Oak Functions has a trusted runtime compiled to a server
```

binary that, for each incoming client gRPC request, executes an untrusted
workload that operates on the request payload, and produces a response which is
sent back to the same client. The trusted runtime ensures that the untrusted
workload may not violate the confidentiality of the client request data,
preventing observers from learning anything about the request. From the client
point of view, the server provides cryptographic evidence (backed by an
hardware-based Trusted Execution Environment) of its own identity as part of a
remote attestation protocol, which convinces the client that it is in fact a
legitimate version of such a runtime.

The main building blocks used in Oak Functions are:

- A Trusted Execution Environment (TEE)

>   - protects the confidentiality and integrity of data and code from the host,

Reviewer: ```suggestion
  - to protect the confidentiality and integrity of data and code from the host,
```
nit: I think it would read better.

Author: Most of the other points use third person, so I kept it here too for now, but I agree it should be made more uniform.

>   - protects the confidentiality and integrity of data and code from the host,

Author: Done.

is responsible for starting the Oak Functions trusted runtime, and loading an
application, with a single Wasm module as the workload. An application may
specify additional resources in a server configuration file. These resources are
as well loaded by the Oak Functions loader at startup time. The trusted runtime
instantiates the given Wasm module for each incoming user request, runs the
request through it in order to produce a response, and then terminates the Wasm

> instance; each Wasm instance is short lived and does not persist state outside

Author: Done.

### ABI Functions and the Rust SDK

WebAssembly is very restricted. In order to be able to implement meaningful
applications in a secure way, the trusted runtime provides a limited number of
ABI functions to the Wasm module. The ABI functions allow reading the incoming
request, writing the response, writing log messages, and fetching items from an
in-memory storage. In addition, there are experimental features for publishing
metrics, and machine-learning inference using a TensorFlow model. The full
description of the ABI functions can be found
[here](https://github.com/project-oak/oak/blob/main/docs/oak_functions_abi.md).

A Rust SDK is currently available, and support for other languages may be added
in the future.

### Read-Only Storage

The trusted runtime does not allow the Wasm workload to interact with any
external resources. However, each Oak Functions application may have read-only
access to an in-memory storage that is populated with data from an external data
source at startup time. The external resource is specified in the server

> configuration file as a URL. It is also possible to specify a refresh interval.

Reviewer: We now also allow reading from file. 

```suggestion
configuration file as a URL or a file path. It is also possible to specify a refresh interval.
```

Author: Done.

This allows the Oak Functions loader to periodically refresh the data in the

> storage, by re-downloading the entirety of the data from the provided URL.

Reviewer: ```suggestion
storage, by re-downloading the entirety of the data from the provided URL or file path.
```

Author: Done.
